docs/downloader.rst

```diff
@@ -0,0 +1,169 @@
+
+This describes the general operation of the remote to local synchronization as outlined in "the Leif Design" (see :ref:`Leif's Proposal: Magic-Folder "single-file" snapshot design`).
+
+We do not describe the synchronization of the list of participants in a magic-folder.
+This list will consist of at least:
+
+- an arbitrary ``name`` for each participant
+- a directory read-capability for each participant ("Personal DMD")
+- in this directory is a flattened representation of every file in  the Magic Folder (similar to the Tahoe 1.14.0 design) with each entry pointing to a Snapshot.
+  That is, there are no sub-directories to recurse into: all files are listed by downloading just the "personal DMD" capability of that participant.
+
+"A Snapshot" is a single version of a single file.
+It is represented by an immutable directory and contains:
+
+- ``content``: (optional) a read-only Capability to the actual content of this Snapshot.
+  Deletion Snapshots contain no such key.
+- ``metadata``: information about the Snapshot, a read-only Capability pointing to a JSON-serialized dict containing:
+  - ``snapshot_version``: int(1) currently
+  - ``name``: the name of this snapshot (a mangled relative path)
+  - ``author``: a dict containing:
+    - ``name``: arbitrary name
+    - ``verify_key``: base64-encoded public key of the author
+  - ``parents``: a list of immutable Capability-strings of any parent Snapshots
+  - (not yet) additionally, in the Tahoe metadata for this metadata-capability is a ``magic_folder`` dict with the following keys:
+    - ``author_signature``: base64-encoded signature which signs the content-Capability, metadata-Capability and name
+
+If there are zero parents this is the first version of a file.
+If there is one parent it is a modification.
+If there are two or more parents this version is the resolution of a conflict.
+
+When reading the code, there is a method ``create_snapshot_from_capability`` which downloads a capability-string from Tahoe and returns a ``RemoteSnapshot`` instance.
+
+
+General Operation
+-----------------
+
+The ``DownloaderService`` is responsible for deciding which Snapshots to download.
+The capability strings of Snapshots to download are given to ``RemoteSnapshotCacheService`` which downloads a Snapshot and all its relevant parents into memory.
+Once a Snapshot is downloaded it is passed to the ``MagicFolderUpdaterService`` which stages the actual content and determins if the change is an "overwrite" or a "conflict" and changes the local filesystem accordingly.
+
+Downloading new Snapshots from other participants ultimately causes changes to the local filesystem (in the magic-folder).
+
+
+What Snapshots to Download
+--------------------------
+
+Each configured magic-folder has a ``collective_dircap`` which is a Tahoe-LAFS Directory Capability ("dircap") for the list of participants.
```

Comment:
Not really: we have a single method of describing the participants in a magic-folder right now: via the single Collective for that magic-folder. If we add more ways, then certainly this may need adjustment at that point.

Reply:
Well, that depends on the resolution of the issue.